lib/ocpp/v201/charge_point.cpp

```diff
@@ -1982,6 +1997,88 @@ void ChargePoint::set_evse_connectors_unavailable(EvseInterface& evse, bool pers
     }
 }
 
+std::optional<ConnectorEnum> ChargePoint::get_evse_connector_type(const uint32_t evse_id, const uint32_t connector_id) {
+    EvseInterface* evse = nullptr;
+    try {
+        evse = &this->evse_manager->get_evse(static_cast<int32_t>(evse_id));
+    } catch (EvseOutOfRangeException&) {
+        return std::nullopt;
+    }
+
+    auto connector = evse->get_connector(static_cast<int32_t>(connector_id));
+    if (connector == nullptr) {
+        return std::nullopt;
+    }
+
+    ComponentVariable connector_cv =
+        ConnectorComponentVariables::get_component_variable(evse_id, connector_id, ConnectorComponentVariables::Type);
+
+    const std::optional<std::string> connector_type =
+        this->device_model->get_optional_value<std::string>(connector_cv, AttributeEnum::Actual);
+    if (!connector_type.has_value()) {
+        return std::nullopt;
+    }
+
+    return conversions::string_to_connector_enum(connector_type.value());
+}
+
+std::optional<ConnectorStatusEnum>
+ChargePoint::get_available_connector_or_status(const uint32_t evse_id, std::optional<ConnectorEnum> connector_type) {
+    EvseInterface* evse;
+    try {
+        evse = &evse_manager->get_evse(static_cast<int32_t>(evse_id));
+    } catch (const EvseOutOfRangeException&) {
+        EVLOG_error << "Evse id " << evse_id << " is not a valid evse id.";
+        return std::nullopt;
+    }
+
+    bool type_found = false;
```

Comment:
I would move the implementation of this to the `Evse` class, minus the evse_id argument ofcourse.